### DISCUSSION TITLE
Tautomer insensitive SMARTS matching

### DISCUSSION THREAD

**Discussion by KKLagoda** [rdkit/rdkit#6507] [Q&A]
Dear community,

I am trying to match a pattern of a molecular substructure to a library of molecules. The pattern is provided in SMARTS, the molecules are provided as SMILES. 
# Expectation
Match the provided SMARTS to the provided SMILES. This requires tautomerisation.
# Setup
```
import rdkit
from rdkit import Chem

smiles = 'COC1=C2C3=C(CNCC[C@]34C=CC(=O)C[C@@H]4O2)C=C1'
smarts = '[#8&v2&R:11]1-[#6:12]-[#6:13]2=[#6:14]-[#6&H2:15]-[#6,#7&H1&v3:16]-[#6&H2:17]-[#6&H2:22]-[#6:21]-2-[#6&H1:18]-1-[#6&H1:23]=[#6:19]-[O&!R:20]'
mol = Chem.MolFromSmiles(smiles)
query = Chem.MolFromSmarts(smarts)
```
### `mol`
![image](https://github.com/rdkit/rdkit/assets/103106752/8e800b44-5509-479b-9339-316265eaa826)
### `query`
![image](https://github.com/rdkit/rdkit/assets/103106752/93efac8d-b887-4d33-9cc7-f51415389728)
# Attempt 1
Inspired by [this blogpost](https://greglandrum.github.io/rdkit-blog/posts/2021-08-03-generalized-substructure-search.html) about generalised substructure search I tried the TautomerQuery:
```
from rdkit.Chem import rdTautomerQuery

tqry = rdTautomerQuery.TautomerQuery(query)
mol.GetSubstructMatches(tqry)
```

which however results in an `ArgumentError`, presumably during the `.GetSubstructMatches()` call. Reversing `mol` and `tqry` alleviates the error, but does not yield a match. Using the magic `query = Chem.AdjustQueryProperties(query)` as mentioned in other discussions did not influence the above results.

## Question 1
_Can I use `TautomerQuery()` in the above scenario? How would I need to modify my code?_
# Attempt 2
Disheartened I turned to `TautomerEnumerator()` to find tautomers for my pattern, which also worked without an error, yet again not delivering the expected result, for any query/mol combination:
```
from rdkit.Chem import MolStandardize

enum = MolStandardize.rdMolStandardize.TautomerEnumerator()
matches = [(
        tautomer,
        qry,
        len(tautomer.GetSubstructMatches(qry))
        )
    for tautomer in enum.Enumerate(mol)
    for qry in enum.Enumerate(query)
]
```
I found the following to work:
```
import functools as ft

def Kekulize(mol, **kwargs):
    Chem.Kekulize(mol, **kwargs)
    return mol

removeAromaticity = ft.partial(Kekulize, clearAromaticFlags=True)
matches = [(
        tautomer,
        qry,
        len(removeAromaticity(tautomer).GetSubstructMatches(qry))
        )
    for tautomer in enum.Enumerate(mol)
    for qry in enum.Enumerate(query)
    ]
```
Allowing me to match `query` correctly by tautomerising `mol`, however wrongly finding a match in case `matches[0]` and unexpectedly not matching in case `matches[2]`. Seemingly the tautomerisation of `query` has no effect on the outcome of the matching.
This time using  `query = Chem.AdjustQueryProperties(query)` leads to only negative results again.
## Question 2
_How can I adjust above example to tautomerise my query?_
(Only relevant if Q1 cannot be implemented)

**Comment by rogersayle** [comment]
This is related to the answer to #6564.  If you wish to support "benzofusion" to your query, then any single bond in the query needs to match single or aromatic, and likewise any double bond in the query needs to match double or aromatic.  This is required to match the bonds between C:12 and C:13 and between C:13 and C:14 against mol.
Hence
`smarts = '[#8&v2&R:11]1-[#6:12]-[#6:13]2=[#6:14]-[#6&H2:15]-[#6,#7&H1&v3:16]-[#6&H2:17]-[#6&H2:22]-[#6:21]-2-[#6&H1:18]-1-[#6&H1:23]=[#6:19]-[O&!R:20]'`
becomes
`smarts = '[#8&v2&R:11]1-[#6:12][#6:13]2=,:[#6:14]-[#6&H2:15]-[#6,#7&H1&v3:16]-[#6&H2:17]-[#6&H2:22]-[#6:21]-2-[#6&H1:18]-1-[#6&H1:23]=[#6:19]-[O&!R:20]'`
or more completely
`smarts = '[#8&v2&R:11]1[#6:12][#6:13]2=,:[#6:14][#6&H2:15][#6,#7&H1&v3:16][#6&H2:17][#6&H2:22][#6:21]2[#6&H1:18]1[#6&H1:23]=,:[#6:19][O&!R:20]'`
